apps/HelloAndroidCamera2/jni/native.cpp

```diff
@@ -0,0 +1,338 @@
+        LOGE("ANativeWindow buffer locked but its format was not YV12.");
+        ANativeWindow_unlockAndPost(win);
+        ANativeWindow_release(win);
+        return false;
+    }
+
+    if (!checkBufferFormatsMatch(srcWidth, srcHeight, srcLumaRowStrideBytes,
+                                 &buf)) {
+        LOGE("ANativeWindow buffer locked but its size was %d x %d, stride = %d"
+             " expected %d x %d, stride = %d",
+             buf.width, buf.height,
+             buf.stride, srcWidth, srcHeight, srcLumaRowStrideBytes);
+        ANativeWindow_unlockAndPost(win);
+        ANativeWindow_release(win);
+        return false;
+    }
+
+    uint8_t *srcLumaPtr = reinterpret_cast<uint8_t *>(
+        env->GetDirectBufferAddress(srcLumaByteBuffer));
+    if (srcLumaPtr == NULL) {
+        return false;
+    }
+
+    uint8_t *dstLumaPtr = reinterpret_cast<uint8_t *>(buf.bits);
+    if (dstLumaPtr == NULL) {
+        return false;
+    }
+
+    uint32_t dstLumaSizeBytes = buf.stride * buf.height;
+    uint32_t dstChromaRowStrideBytes = ALIGN(buf.stride / 2, 16);
+    // Size of one chroma plane.
+    uint32_t dstChromaSizeBytes = dstChromaRowStrideBytes * buf.height / 2;
+    uint8_t *dstChromaVPtr = dstLumaPtr + dstLumaSizeBytes;
+    uint8_t *dstChromaUPtr = dstLumaPtr + dstLumaSizeBytes + dstChromaSizeBytes;
+
+// If we're using opencl, use the gpu backend for it.
+#if COMPILING_FOR_OPENCL
+    halide_opencl_set_device_type("gpu");
+#endif
+
+    // Make these static so that we can reuse device allocations across frames.
+    static buffer_t srcBuf = { 0 };
+    static buffer_t dstBuf = { 0 };
+
+    srcBuf.host = srcLumaPtr;
+    srcBuf.host_dirty = true;
+    srcBuf.extent[0] = srcWidth;
+    srcBuf.extent[1] = srcHeight;
+    srcBuf.extent[2] = 0;
+    srcBuf.extent[3] = 0;
+    srcBuf.stride[0] = 1;
+    srcBuf.stride[1] = srcLumaRowStrideBytes;
+    srcBuf.min[0] = 0;
+    srcBuf.min[1] = 0;
+    srcBuf.elem_size = 1;
+
+    dstBuf.host = dstLumaPtr;
+    dstBuf.extent[0] = buf.width;  // src and dst width/height actually match.
+    dstBuf.extent[1] = buf.height;
+    dstBuf.extent[2] = 0;
+    dstBuf.extent[3] = 0;
+    dstBuf.stride[0] = 1;
+    dstBuf.stride[1] = buf.stride;  // src and dst strides actually match.
+    dstBuf.min[0] = 0;
+    dstBuf.min[1] = 0;
+    dstBuf.elem_size = 1;
+
+    static bool first_call = true;
+    static unsigned counter = 0;
+    static unsigned times[16];
+    if (first_call) {
+        LOGD("According to Halide, host system has %d cpus\n",
+             halide_host_cpu_count());
+        first_call = false;
+        for (int t = 0; t < 16; t++) {
+            times[t] = 0;
+        }
+    }
+
+    // Set chrominance to 128 to appear grayscale.
+    // The dst chroma is guaranteed to be tightly packed since it's YV12.
+    memset(dstChromaVPtr, 128, dstChromaSizeBytes * 2);
+
+    int64_t t1 = halide_current_time_ns();
+    edge_detect(&srcBuf, &dstBuf);
```

Comment:
halide functions return an error code; it's good form to always check the result. (if you override halide_error() to do something like abort() or exit() then this isn't necessary, of course, since the function would never return a nonzero error...)